optimism/src/trace.rs

```diff
@@ -0,0 +1,62 @@
+use crate::{lookups::Lookup, E};
+use ark_ff::Field;
+use kimchi_msm::witness::Witness;
+use std::{collections::HashMap, hash::Hash};
+
+pub struct Trace<const N: usize, SELECTOR, F> {
+    /// The domain size of the circuit
+    pub domain_size: usize,
+    /// The witness for a given selector
+    pub witness: HashMap<SELECTOR, Witness<N, Vec<F>>>,
```

Comment:
Note: if you want to process in order the keys (and therefore the witness), you need to use `BTreeMap`. `keys` and `values` are not deterministic on `HashMap`, but are on `BTreeMap`

Reply:
For now I don't care about the order because the main is creating one proof for each different witness type. But I agree with you that when real folding is involved and there's just one single final proof, then yes BTreeMap will be a better structure. But for now I wanted to keep it simple as hash maps, because this struct will surely be subject to change as soon as I try to integrate the interface for folding.